extensions/stack_snapshot.bzl

```diff
@@ -0,0 +1,295 @@
+            """,
+        ),
+        "hidden": attr.bool(
+            default = False,
+            doc = "Configuring a package makes it visible by default, unless `hidden` is set to True",
+        ),
+    },
+)
+
+def _assert_unique_tag(tags, tag_name, module):
+    """ Errors if the module used the `tag_name` tag more than once"""
+    if len(tags) > 1:
+        fail(
+            """Module "{module_name}~{module_version}" used configuration tag "{tag_name}" more than once""".format(
+                module_name = module.name,
+                module_version = module.version,
+                tag_name = tag_name,
+            ),
+        )
+
+def _add_packages(conf, module, root_or_rules_haskell):
+    """Read the `package` tags from `module` and add the configuration to `conf`"""
+    packages_in_module = sets.make()
+    for package_tag in module.tags.package:
+        package_name = package_tag.name
+
+        # Check that a package is configure at most one time per module
+        if sets.contains(packages_in_module, package_name):
+            fail("""Module "{module_name}~{module_version}" configured package "{package_name}" multiple times.""".format(
+                module_version = module.version,
+                module_name = module.name,
+                package_name = package_name,
+            ))
+        sets.insert(packages_in_module, package_name)
+
+        # If the package is already configured, we can only only make it visible.
+        if sets.contains(conf.configured_packages, package_name):
+            if not package_tag.hidden and package_name not in conf.vendored_packages:
+                sets.insert(conf.packages, package_name)
+            continue
+        sets.insert(conf.configured_packages, package_name)
+
+        # We are the first one to configure the package:
+        if not package_tag.hidden and not package_tag.vendored:
+            sets.insert(conf.packages, package_name)
+        if package_tag.components != ["DETECT_DEFAULT"]:
+            # Some packages have default components set
+            conf.components[package_name] = package_tag.components
+        if package_tag.components_dependencies:
+            conf.components_dependencies[package_name] = json.encode(package_tag.components_dependencies)
+        if package_tag.extra_deps:
+            conf.extra_deps[package_name] = package_tag.extra_deps
+        if root_or_rules_haskell:
+            # Only the root module or rules_haskell can modify the
+            # following configuration settings. To make the configuration
+            # easier to understand.
+            if package_tag.setup_deps:
+                conf.setup_deps[package_name] = package_tag.setup_deps
+            if package_tag.flags:
+                conf.flags[package_name] = package_tag.flags
+            if package_tag.vendored:
+                conf.vendored_packages[package_name] = package_tag.vendored
```

Comment:
I'm happy to keep it this way in this PR. For a follow-up it's worth considering whether these should raise an error on non-root/non-rules_haskell modules. Silently ignoring tag properties could be quite confusing to unsuspecting users.